### PR TITLE
Feat/new rate setters

### LINKED ISSUE #2322: Deficit-based rate setter
Implement a rate setter based only on deficit and estimate of its growth. See the [Red Giant summary document](https://docs.google.com/document/d/16UoHdCKx-EjzsC5Yet6AVmk9-3tjO3abZJ3HdgArmBQ/edit#heading=h.893k5mtqz8y) for details of the proposal.

### PR DESCRIPTION
**Pull request by cyberphysic4l**
# Description of change

Implementation of three rate setters: deficit (default), AIMD and disabled.

## Type of change

Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Run on docker network. Still some issues after running for a few minutes with spammer activated.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] My code follows the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -347,3 +347,51 @@ jobs:
         with:
           name: ${{ env.TEST_NAME }}
           path: tools/integration-tests/logs
+
+  orphanage:
```
Comment (piotrm50) [inline .github/workflows/integration-tests.yml]: Isn't this part of develop already?

File: deploy/ansible/roles/goshimmer-node/templates/docker-compose-entrynode.yml.j2 [hunk 2]
```diff
@@ -7,7 +7,8 @@ services:
     container_name: entrynode
     restart: unless-stopped
     volumes:
-      - "./db:/db"
+      - "./db:/app/db"
```
Comment (piotrm50) [inline deploy/ansible/roles/goshimmer-node/templates/docker-compose-entrynode.yml.j2]: Isn't this on develop already as well?

File: tools/docker-network/run.sh [hunk 3]
```diff
@@ -46,4 +46,4 @@ export COMPOSE_PROFILES=$(join , ${PROFILES[@]})
 docker compose up
 
 echo "Clean up docker resources"
-docker compose down -v
+docker compose down -v
```
Comment (piotrm50) [inline tools/docker-network/run.sh]: This should not change on this PR

File: tools/integration-tests/tester/framework/parameters.go [hunk 4]
```diff
@@ -89,7 +89,6 @@ func PeerConfig() config.GoShimmer {
 	c.Notarization.MinEpochCommittableAge = 10 * time.Second
 
 	c.BlockIssuer.Enabled = true
-	c.BlockIssuer.RateSetter.Enable = false
```
Comment (piotrm50) [inline tools/integration-tests/tester/framework/parameters.go:89]: Do we still disable rate setter in integration tests somehow?

File: tools/docker-network/docker-compose.yml [hunk 5]
```diff
@@ -133,7 +133,7 @@ services:
       - faucet
 
   faucet:
-    image: docker-network-peer_master:latest
+    image: docker-network_peer_master:latest
```
Comment (piotrm50) [inline tools/docker-network/docker-compose.yml]: Do not change this.

File: tools/docker-network/docker-compose.yml [hunk 6]
```diff
@@ -73,7 +73,7 @@ services:
     depends_on:
       - peer_master
   peer_master2:
-    image: docker-network-peer_master:latest
+    image: docker-network_peer_master:latest
```
Comment (piotrm50) [inline tools/docker-network/docker-compose.yml]: Do not change this.

File: tools/docker-network/docker-compose.yml [hunk 7]
```diff
@@ -48,7 +48,7 @@ services:
       - goshimmer
 
   peer_replica:
-    image: docker-network-peer_master:latest
+    image: docker-network_peer_master:latest
```
Comment (piotrm50) [inline tools/docker-network/docker-compose.yml]: Do not change this.

File: tools/docker-network/docker-compose.yml [hunk 8]
```diff
@@ -84,7 +84,7 @@ services:
       --node.peerDBDirectory=/app/peerdb
       --database.directory=/app/db
       --protocol.snapshot.path=./snapshot.bin
-      --blockIssuer.rateSetter.enable=false
+      --blockIssuer.rateSetter.mode=deficit
```
Comment (piotrm50) [inline tools/docker-network/docker-compose.yml]: On some nodes we disable ratesetter and on some we use `deficit`. This should be uniform.

File: tools/docker-network/config.docker.json [hunk 9]
```diff
@@ -58,7 +58,7 @@
   },
   "node": {
     "disablePlugins": "portcheck",
-    "enablePlugins": []
+    "enablePlugins": "Spammer"
```
Comment (piotrm50) [inline tools/docker-network/config.docker.json]: It's better to enable plugins in the docker-compose file, because it's more intuitive if we to disable them or check what is enabled, we just need to look there.

File: plugins/blockissuer/parameters.go [hunk 10]
```diff
@@ -10,12 +10,12 @@ import (
 type ParametersDefinition struct {
 	// RateSetter contains the definition of the parameters used by the Rate Setter.
 	RateSetter struct {
+		// Mode determines the type of the rate setting mechanism.
+		Mode string `default:"deficit" usage:"define the type of rate setter to use"`
```
Comment (piotrm50) [inline plugins/blockissuer/parameters.go]: List possible rate setter options  in the usage desccription.

File: plugins/activity/plugin.go [hunk 11]
```diff
@@ -39,15 +39,15 @@ func configure(plugin *node.Plugin) {
...
+	for {
+		if estimate := deps.BlockIssuer.Estimate(); estimate > 0 {
+			time.Sleep(estimate)
+		} else {
+			break
+		}
 	}
-	err = deps.BlockIssuer.IssueBlockAndAwaitBlockToBeScheduled(block, Parameters.BroadcastInterval)
+
+	block, err := deps.BlockIssuer.IssuePayload(activityPayload, Parameters.ParentsCount)
```
Comment (piotrm50) [inline plugins/activity/plugin.go]: Something merged wrong.

```suggestion
	block, err := deps.BlockIssuer.CreateBlock(activityPayload, Parameters.ParentsCount)
	if err != nil {
		Plugin.LogWarnf("error creating activity block: %s", err)
		return
	}
	err = deps.BlockIssuer.IssueBlockAndAwaitBlockToBeScheduled(block, Parameters.BroadcastInterval)
```

File: packages/protocol/testframework.go [hunk 12]
```diff
@@ -4,6 +4,8 @@ import (
 	"os"
 	"testing"
 
+	"github.com/iotaledger/goshimmer/packages/protocol/congestioncontrol"
```
Comment (piotrm50) [inline packages/protocol/testframework.go]: Is this import correctly sorted?

File: packages/protocol/settings.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/iotaledger/hive.go/core/generics/set"
+	"github.com/iotaledger/hive.go/core/serix"
+
+	"github.com/iotaledger/goshimmer/packages/core/commitment"
+	"github.com/iotaledger/goshimmer/packages/core/storable"
+)
+
+// region Settings /////////////////////////////////////////////////////////////////////////////////////////////////////
+
+type Settings struct {
```
Comment (piotrm50) [inline packages/protocol/settings.go]: Seems like this file should not be part of this PR.

File: packages/protocol/settings_test.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/iotaledger/goshimmer/packages/core/commitment"
+)
+
+func TestSettings(t *testing.T) {
```
Comment (piotrm50) [inline packages/protocol/settings_test.go]: Seems like this file should not be part of this PR.

File: packages/protocol/protocol.go [hunk 15]
```diff
@@ -210,16 +210,15 @@ func (p *Protocol) initTipManager() {
 	p.Events.TipManager = p.TipManager.Events
 }
 
-func (p *Protocol) ProcessBlock(block *models.Block, src identity.ID) {
+func (p *Protocol) ProcessBlock(block *models.Block, src identity.ID) error {
 	isSolid, chain, _ := p.chainManager.ProcessCommitment(block.Commitment())
 	if !isSolid {
-		fmt.Println(">> chain not solid", block.Commitment().ID(), "latest commitment", p.storage.Settings.LatestCommitment().ID(), "blockID", block.ID())
-		return
+		return errors.Errorf(">> chain not solid", block.Commitment().ID(), "latest commitment", p.storage.Settings.LatestCommitment().ID(), "blockID", block.ID())
```
Comment (piotrm50) [inline packages/protocol/protocol.go]: Error message should be better formulated. This was just a debugging print that was intended to be removed.

File: packages/protocol/engine/notarization/proofs/proofs.go [hunk 16]
```diff
@@ -71,7 +83,11 @@ func (f *commitmentFactory) ProofStateRoot(ei epoch.Index, outID utxo.OutputID)
 }
 
 // ProofStateMutationRoot returns the merkle proof for the transactionID against the state mutation root.
+<<<<<<<< HEAD:packages/protocol/engine/notarization/proofs.go
```
Comment (piotrm50) [inline packages/protocol/engine/notarization/proofs/proofs.go]: Looks like this file contains some unresolved merge conflicts.

File: packages/protocol/engine/notarization/epochmutations.go [hunk 17]
```diff
@@ -154,7 +154,6 @@ func (m *EpochMutations) Evict(index epoch.Index) (acceptedBlocks *ads.Set[model
 
 	return m.acceptedBlocks(index), m.acceptedTransactions(index), lo.Return1(m.attestationsByEpoch.Get(index)), nil
 }
-
 // Attestations returns the attestations for the given epoch.
```
Comment (piotrm50) [inline packages/protocol/engine/notarization/epochmutations.go]:
```suggestion

// Attestations returns the attestations for the given epoch.
```

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 18]
```diff
@@ -57,11 +54,13 @@ type Scheduler struct {
 	optsRate                           time.Duration
 	optsMaxBufferSize                  int
 	optsAcceptedBlockScheduleThreshold time.Duration
+	optsMaxDeficit                     *big.Rat
 
 	started        typeutils.AtomicBool
 	stopped        typeutils.AtomicBool
 	shutdownSignal chan struct{}
 	shutdownOnce   sync.Once
+	readyChan      chan *Block
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: Is this channel used anywhere?

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 19]
```diff
@@ -412,13 +434,12 @@ loop:
 		select {
 		// every rate time units
 		case <-s.ticker.C:
-			// TODO: pause the ticker, if there are no ready blocks
 			if block := s.schedule(); block != nil {
+				//s.ticker.Reset(s.optsRate * time.Duration(block.Work()))
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: What is this? Also, the TODO above is removed, and it's still not done :smiley:

File: packages/app/blockissuer/blockissuer.go [hunk 20]
```diff
@@ -27,12 +27,12 @@ type BlockIssuer struct {
...
 
 	*blockfactory.Factory
-	*ratesetter.RateSetter
+	ratesetter.RateSetter
 	protocol          *protocol.Protocol
 	identity          *identity.LocalIdentity
 	referenceProvider *blockfactory.ReferenceProvider
 
-	optsRateSetterOptions      []options.Option[ratesetter.RateSetter]
+	optsRateSetterMode         ratesetter.ModeType
```
Comment (piotrm50) [inline packages/app/blockissuer/blockissuer.go]: This value is not used anywhere, so it should be removed.

File: packages/app/blockissuer/ratesetter/deficit/ratesetter.go [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+	events                *utils.Events
+	protocol              *protocol.Protocol
+	self                  identity.ID
+	totalManaRetrieveFunc func() int64
+	manaRetrieveFunc      func() map[identity.ID]int64
+
+	ownRate           *atomic.Float64
+	maxRate           float64
+	initOnce          sync.Once
+	deficitsMutex     sync.RWMutex
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/deficit/ratesetter.go]: don't need this mutex

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 22]
```diff
@@ -1,88 +1,78 @@
...
+
+	"github.com/iotaledger/hive.go/core/generics/event"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRateSetter_IssueBlockAndAwaitSchedule(t *testing.T) {
+	allModes := []ModeType{AIMDMode, DeficitMode, DisabledMode}
+
+	for _, mode := range allModes {
+		func() {
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]: Why have a function here?

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 23]
```diff
@@ -1,88 +1,78 @@
...
+func TestRateSetter_IssueBlocksAndAwaitScheduleMultipleIssuers(t *testing.T) {
+	allModes := []ModeType{DeficitMode, DisabledMode}
+	numIssuers := 5
+	numBlocksPerIssuer := 10
+	schedulerRate := 100
+
+	allBlocks := make(map[models.BlockID]*models.Block)
+
+	for _, mode := range allModes {
+		func() {
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]: Why have a function here?

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 24]
```diff
@@ -1,88 +1,78 @@
...
+			tf := NewTestFramework(t, WithRateSetterOptions(WithMode(mode)), WithSchedulerOptions(scheduler.WithRate(time.Duration(schedulerRate))), WithNumIssuers(numIssuers))
+			blockScheduled := make(chan *models.Block, 1)
+			tf.Protocol.CongestionControl.Scheduler().Events.BlockScheduled.Attach(event.NewClosure(func(block *scheduler.Block) { blockScheduled <- block.ModelsBlock }))
+
+			for i := 0; i < numIssuers; i++ {
+				defer tf.RateSetter[i].Shutdown()
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]: `Shutdown` should be member of the testframework, and `shutdown tf.Shutdown()` should be put right after testframework creation.

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 25]
```diff
@@ -1,88 +1,78 @@
...
+	allModes := []ModeType{DeficitMode, DisabledMode}
+	numIssuers := 5
+	numBlocksPerIssuer := 10
+	schedulerRate := 100
+
+	allBlocks := make(map[models.BlockID]*models.Block)
+
+	for _, mode := range allModes {
+		func() {
+			tf := NewTestFramework(t, WithRateSetterOptions(WithMode(mode)), WithSchedulerOptions(scheduler.WithRate(time.Duration(schedulerRate))), WithNumIssuers(numIssuers))
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]:
```suggestion
			tf := NewTestFramework(t, WithRateSetterOptions(WithMode(mode)), WithSchedulerOptions(scheduler.WithRate(schedulerRate)), WithNumIssuers(numIssuers))
```

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 26]
```diff
@@ -1,88 +1,78 @@
...
+			}, 1*time.Second, 10*time.Millisecond)
+		}()
+	}
+}
+
+func TestRateSetter_IssueBlocksAndAwaitScheduleMultipleIssuers(t *testing.T) {
+	allModes := []ModeType{DeficitMode, DisabledMode}
+	numIssuers := 5
+	numBlocksPerIssuer := 10
+	schedulerRate := 100
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]:
```suggestion
	schedulerRate := time.Duration(1)*time.Millisecond
```

File: packages/app/blockissuer/ratesetter/ratesetter_test.go [hunk 27]
```diff
@@ -1,88 +1,78 @@
...
+						delete(allBlocks, blk.ID())
+						if len(allBlocks) == 0 {
+							return true
+						}
+					default:
+						return false
+					}
+				}
+			}, 1*time.Second, 10*time.Millisecond)
+		}()
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/ratesetter_test.go]:
```suggestion
			assert.Eventually(t, func() bool {
			        select {
				case blk := <-blockScheduled:
					delete(allBlocks, blk.ID())
					return len(allBlocks) == 0
				default:
					return false
				}
			}, 1*time.Second, 10*time.Millisecond)
		}()
```

File: packages/app/blockissuer/ratesetter/testframework.go [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+	for {
+		if estimate := tf.RateSetter[issuer].Estimate(); estimate > 0 {
+			time.Sleep(estimate)
+		} else {
+			tf.Protocol.ProcessBlock(block, tf.localIdentity[issuer].ID())
+			break
+		}
+	}
+	return nil
+}
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/testframework.go]:
```suggestion
func (tf *TestFramework) IssueBlock(block *models.Block, issuer int) error {
	for estimate := tf.RateSetter[issuer].Estimate(); estimate > 0; estimate = tf.RateSetter[issuer].Estimate() {
		time.Sleep(estimate)
	}

	return tf.Protocol.ProcessBlock(block, tf.localIdentity[issuer].ID())
}
```

File: packages/app/blockissuer/ratesetter/utils/errors.go [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+
+var (
+	// ErrInvalidIssuer is returned when an invalid block is passed to the rate setter.
+	ErrInvalidIssuer = errors.New("block not issued by local node")
+
+	// ErrBlockWasNotScheduledInTime is returned if a block did not get issued within the defined await time.
+	ErrBlockWasNotScheduledInTime = errors.New("block could not be scheduled in time")
+
+	// ErrStopped is returned when a block is passed to a stopped rate setter.
+	ErrStopped = errors.New("rate setter stopped")
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/utils/errors.go]: Those variables are not used, so the entire file should be removed.

File: packages/app/blockissuer/ratesetter/utils/events.go [hunk 30]
```diff
@@ -8,14 +8,12 @@ import (
 
 type Events struct {
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/utils/events.go]: Events not used anywhere, so remove the file.

File: packages/app/blockissuer/ratesetter/utils/issuerqueue.go [hunk 31]
```diff
@@ -10,20 +10,26 @@ import (
...
 
+const (
+	// MaxLocalQueueSize is the maximum local (containing the block to be issued) queue size in number of blocks.
+	MaxLocalQueueSize = 20
+)
+
 // region IssuerQueue /////////////////////////////////////////////////////////////////////////////////////////////
 
 // IssuerQueue keeps the submitted blocks of an issuer.
 type IssuerQueue struct {
```
Comment (piotrm50) [inline packages/app/blockissuer/ratesetter/utils/issuerqueue.go]: IssuerQueue is not used anymore in the ratesetters, so remove the whole file.

File: packages/app/spammer/spammer.go [hunk 32]
```diff
@@ -83,9 +83,14 @@ func (s *Spammer) run(rate int, payloadSize uint64, timeUnit time.Duration, imif
...
-			estimatedDuration := s.estimateFunc()
 			// TODO: only sleep if estimate > some threshold.
-			time.Sleep(estimatedDuration)
+			for {
+				if estimatedDuration := s.estimateFunc(); estimatedDuration > 0 {
+					time.Sleep(estimatedDuration)
+				} else {
+					break
+				}
+			}
```
Comment (piotrm50) [inline packages/app/spammer/spammer.go]: Would the following be cleaner?

```suggestion
			for estimatedDuration := s.estimateFunc(); estimatedDuration > 0; estimatedDuration = s.estimateFunc() {
				time.Sleep(estimatedDuration)
			}
```

File: packages/network/warpsync/validation.go [hunk 33]
```diff
@@ -0,0 +1,14 @@
+package warpsync
+
+import (
+	"github.com/iotaledger/hive.go/core/identity"
+
+	"github.com/iotaledger/goshimmer/packages/core/epoch"
+)
+
+func (p *Protocol) RequestEpochCommittment(ei epoch.Index, to ...identity.ID) {
```
Comment (piotrm50) [inline packages/network/warpsync/validation.go]: This file should not be modified, so it's most likely a result of an incorrect merge?

File: packages/protocol/congestioncontrol/icca/scheduler/bufferqueue.go [hunk 34]
```diff
@@ -39,12 +39,12 @@ func (b *BufferQueue) NumActiveIssuers() int {
...
 }
 
-// MaxSize returns the max size (in bytes) of all blocks in b.
+// MaxSize returns the max number of blocks in b.
 func (b *BufferQueue) MaxSize() int {
 	return b.maxBuffer
 }
 
-// Size returns the total size (in bytes) of all blocks in b.
+// Size returns the total number of block in b.
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/bufferqueue.go]:
```suggestion
// Size returns the total number of blocks in BufferQueue.
```

File: packages/protocol/congestioncontrol/icca/scheduler/bufferqueue.go [hunk 35]
```diff
@@ -39,12 +39,12 @@ func (b *BufferQueue) NumActiveIssuers() int {
 	return b.activeIssuers.Size()
 }
 
-// MaxSize returns the max size (in bytes) of all blocks in b.
+// MaxSize returns the max number of blocks in b.
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/bufferqueue.go]:
```suggestion
// MaxSize returns the max number of blocks in BufferQueue.
```

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 36]
```diff
@@ -78,9 +77,11 @@ func New(evictionState *eviction.State, isBlockAccepted func(models.BlockID) boo
 		blocks:                             memstorage.NewEpochStorage[models.BlockID, *Block](),
 		optsMaxBufferSize:                  300,
 		optsAcceptedBlockScheduleThreshold: 5 * time.Minute,
-		optsRate:                           5 * time.Millisecond,
+		optsRate:                           5 * time.Millisecond,      // measured in time per unit work
+		optsMaxDeficit:                     new(big.Rat).SetInt64(10), // must be >= max block work, but work is currently=1 for all blocks
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go:80]: So an issuer can accumulate deficit to issue maximum 10 blocks, is this a reasonable default?

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 37]
```diff
@@ -467,8 +488,13 @@ func (s *Scheduler) schedule() *Block {
...
 	block := s.buffer.PopFront()
 	issuerID := identity.NewID(block.IssuerPublicKey())
-	s.updateDeficit(issuerID, new(big.Rat).SetInt64(-int64(block.Size())))
-
+	deficit := s.Deficit(issuerID)
+	if deficit.Cmp(new(big.Rat).SetInt64(int64(block.Work()))) < 0 {
+		deficitFloat, _ := deficit.Float64()
+		errorString := fmt.Sprintf("scheduler: deficit is less than block work - Deficit is %d, block work is %d, ", int(deficitFloat), block.Work())
+		panic(errorString)
+	}
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: This check is already covered in `updateDeficit` function. 

```suggestion
```

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 38]
```diff
@@ -556,7 +582,20 @@ func (s *Scheduler) updateDeficit(issuerID identity.ID, d *big.Rat) {
...
+func (s *Scheduler) GetExcessDeficit(issuerID identity.ID) (float64, error) {
+	var deficitFloat float64
+	s.deficitsMutex.RLock()
+	defer s.deficitsMutex.RUnlock()
+	if deficit, exists := s.deficits.Get(issuerID); exists {
+		deficitFloat, _ = deficit.Float64()
+	} else {
+		return 0.0, errors.New(fmt.Sprintf("Deficit for issuer %s does not exist", issuerID))
+	}
+	return deficitFloat - float64(s.IssuerQueueWork(issuerID)), nil
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: I think the following is cleaner.

```suggestion
func (s *Scheduler) GetExcessDeficit(issuerID identity.ID) (deficitFloat float64, err error) {
	s.deficitsMutex.RLock()
	defer s.deficitsMutex.RUnlock()

	if deficit, exists := s.deficits.Get(issuerID); !exists {
		deficitFloat, _ = deficit.Float64()
		return deficitFloat - float64(s.IssuerQueueWork(issuerID)), nil
	}

	return 0.0, errors.New(fmt.Sprintf("Deficit for issuer %s does not exist", issuerID))
```

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 39]
```diff
@@ -614,6 +653,12 @@ func WithRate(rate time.Duration) options.Option[Scheduler] {
 	}
 }
 
+func WithMaxDeficit(maxDef int) options.Option[Scheduler] {
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go:666]: This is not used anywhere. Add this parameter to `SchedulerParametersDefinition` and pass this value down.

**Review by piotrm50 — CHANGES_REQUESTED** [review]

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 36 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go:80]: I think it seems reasonable tbh.. If they have enough access mana this deficit will grow very fast anyway. I often used only one or two as max deficit in simulations for this value.

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 37 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: True, this extra check was left over from debugging.

File: packages/app/blockissuer/blockissuer.go (hunk 20 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/blockissuer.go]: done

File: packages/app/blockissuer/ratesetter/utils/issuerqueue.go (hunk 31 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/utils/issuerqueue.go]: done

File: packages/app/blockissuer/ratesetter/ratesetter_test.go (hunk 22 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: I have no idea...

Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: removed

File: packages/app/blockissuer/ratesetter/ratesetter_test.go (hunk 23 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: removed

File: packages/app/blockissuer/ratesetter/ratesetter_test.go (hunk 26 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: Made similar change to be duration but keeping to 100 nanoseconds.

File: packages/app/blockissuer/ratesetter/ratesetter_test.go (hunk 25 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: done

File: packages/app/blockissuer/ratesetter/deficit/ratesetter.go (hunk 21 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/deficit/ratesetter.go]: fair enough

File: plugins/blockissuer/parameters.go (hunk 10 above)
Comment (cyberphysic4l) [reply plugins/blockissuer/parameters.go]: ok

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 19 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: Comment was part of another TODO which was left in accidentally. Essentially we would like to make the schedule work in units of work, not in number of blocks, so this requires a different wait time after each schedule, no a fixed wait time.
As for the removed TODO, you can't pause a ticker, you can only stop it and then it can't be restarted, so some different approach is required for that as well.
I will add a new TODO to capture all changes we would like to make.

Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]:
```suggestion
// TODO: make this operate in units of work, with a variable pause between scheduling depending on work scheduled.
	// TODO: don't use a ticker. Switch to a simple timer instead, and use a flag when ready to schedule something if there is nothing ready to be scheduled yet.
	// TODO: implement a token bucket for the scheduler to account for bursty arrivals.
```

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 18 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: nope, not anymore, removed.

File: packages/protocol/engine/notarization/proofs/proofs.go (hunk 16 above)
Comment (cyberphysic4l) [reply packages/protocol/engine/notarization/proofs/proofs.go]: doh

File: packages/protocol/testframework.go (hunk 12 above)
Comment (cyberphysic4l) [reply packages/protocol/testframework.go]: fixed

File: tools/docker-network/config.docker.json (hunk 9 above)
Comment (cyberphysic4l) [reply tools/docker-network/config.docker.json]: ok, this was left over from when I was having trouble with spammer at one point.

File: tools/integration-tests/tester/framework/parameters.go (hunk 4 above)
Comment (cyberphysic4l) [reply tools/integration-tests/tester/framework/parameters.go:89]: Nope, the default would be used here I guess, which is deficit mode.

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 39 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go:666]: done

File: tools/integration-tests/tester/framework/parameters.go (hunk 4 above)
Comment (cyberphysic4l) [reply tools/integration-tests/tester/framework/parameters.go:89]: Will add Mode=disabled here

File: tools/docker-network/docker-compose.yml (hunk 8 above)
Comment (cyberphysic4l) [reply tools/docker-network/docker-compose.yml]: Delete this line for all nodes

File: packages/app/blockissuer/ratesetter/ratesetter_test.go (hunk 24 above)
Comment (cyberphysic4l) [reply packages/app/blockissuer/ratesetter/ratesetter_test.go]: fixed

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go [hunk 40]
```diff
@@ -566,7 +586,19 @@ func (s *Scheduler) updateDeficit(issuerID identity.ID, d *big.Rat) {
...
+func (s *Scheduler) GetExcessDeficit(issuerID identity.ID) (deficitFloat float64, err error) {
+	s.deficitsMutex.RLock()
+	defer s.deficitsMutex.RUnlock()
+
+	if deficit, exists := s.deficits.Get(issuerID); exists {
+		deficitFloat, _ = deficit.Float64()
+		return deficitFloat - float64(s.IssuerQueueWork(issuerID)), nil
+	}
+
+	return 0.0, errors.New(fmt.Sprintf("Deficit for issuer %s does not exist", issuerID))
```
Comment (piotrm50) [inline packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]:
```suggestion
	return 0.0, errors.Errorf("Deficit for issuer %s does not exist", issuerID)
```

File: packages/app/spammer/spammer.go [hunk 41]
```diff
@@ -83,9 +83,13 @@ func (s *Spammer) run(rate int, payloadSize uint64, timeUnit time.Duration, imif
 		case <-s.shutdown:
 			return
 		case <-ticker.C:
-			estimatedDuration := s.estimateFunc()
 			// TODO: only sleep if estimate > some threshold.
-			time.Sleep(estimatedDuration)
+			for estimatedDuration := s.estimateFunc(); estimatedDuration > 0; estimatedDuration = s.estimateFunc() {
+				if estimatedDuration > time.Duration(rate) {
+					estimatedDuration = time.Duration(rate)
+				}
```
Comment (piotrm50) [inline packages/app/spammer/spammer.go]: Use something like below instead of `if`. Return value of `min` can be also put directly into `time.Sleep`

```
estimatedDuration = lo.Min(estimatedDuration, time.Duration(rate))
```

File: plugins/activity/plugin.go [hunk 42]
```diff
@@ -39,8 +39,13 @@ func configure(plugin *node.Plugin) {
...
 	activityPayload := payload.NewGenericDataPayload([]byte("activity"))
-
-	time.Sleep(deps.BlockIssuer.Estimate())
+	for {
+		if estimate := deps.BlockIssuer.Estimate(); estimate > 0 {
+			time.Sleep(estimate)
+		} else {
+			break
+		}
+	}
```
Comment (piotrm50) [inline plugins/activity/plugin.go:48]: Could we end up in a something like infinite loop here? Should we introduce some value that limits the total time waiting? Or number of times we wait? But on the other hand if it's told to wait all the time, that means that the node is issuing other blocks, so activity blocks are not really necessary. So I think it's safe to leave it as is. WDYT?

**Review by piotrm50 — APPROVED** [review]
I think it's good to go! I added some minor improvement suggestions, but otherwise I think it's good to go!

File: packages/protocol/congestioncontrol/icca/scheduler/scheduler.go (hunk 40 above)
Comment (cyberphysic4l) [reply packages/protocol/congestioncontrol/icca/scheduler/scheduler.go]: done

File: packages/app/spammer/spammer.go (hunk 41 above)
Comment (cyberphysic4l) [reply packages/app/spammer/spammer.go]: done

File: plugins/activity/plugin.go (hunk 42 above)
Comment (cyberphysic4l) [reply plugins/activity/plugin.go:48]: I think this could result in very long sleep times in the case of low mana nodes, but activity blocks should not really be goverened by the congestion control for regular blocks in the long run anyway, they should be part of a separate allocation for validators. I think it is fine to leave it as it is for now.
